[Xcode] Set MACOSX_DEPLOYMENT_TARGET on open source builds to match the system version

emw-apple opened the pull request:
#### 8dbd7e57ce021a18b8942a639b8bef3ceaeb9433
<pre>
[Xcode] Set MACOSX_DEPLOYMENT_TARGET on open source builds to match the system version
<a href="https://bugs.webkit.org/show_bug.cgi?id=252783">https://bugs.webkit.org/show_bug.cgi?id=252783</a>

Reviewed by Alexey Proskuryakov.

While internal builds of WebKit do not support back deployment
(targeting an older OS than what an SDK was built for), it&apos;s possible to
make open source builds in this way. Currently, we assume that the SDK&apos;s
default deployment target is fine, and make no effort to ensure that
what we build is actually runnable on the system.

This feels like unnecessary friction for open source contributors,
especially since the Xcode available in an older OS&apos;s App Store contains
SDKs for newer macOS versions. Rather than building artifacts that won&apos;t
run on the host machine, we should gracefully fall back.

Achieve this in build-webkit by passing MACOSX_DEPLOYMENT_TARGET, based
on the system&apos;s version. The deployment target can be overridden (or
unset) by passing MACOSX_DEPLOYMENT_TARGET on the command line.

See also: <a href="https://commits.webkit.org/245182@main">https://commits.webkit.org/245182@main</a>, which stopped setting
the *_DEPLOYMENT_TARGET settings automatically in order to support
building for minor-release deployment targets.

* Tools/Scripts/build-webkit:

Canonical link: <a href="https://commits.webkit.org/260719@main">https://commits.webkit.org/260719@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ebdf23574365cc43289d968f2c410c81f00bf9fe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109153 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18235 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41970 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/683 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/118398 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113035 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19694 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9537 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101393 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114911 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/97996 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/42938 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29650 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/84670 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/98160 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11014 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30993 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/98953 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/9099 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11744 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/7929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/31031 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17113 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50596 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/106641 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7392 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13361 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26424 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov reviewed (approved):
> But in public Xcode projects, this is safe to do and a common build configuration.

I wouldn't be so categorical about this. We've seen many issues in this configuration, and we don't actively support it. We do however feel that it's a valid configuration.

litherum commented on Tools/Scripts/build-webkit:
```diff
@@ -364,6 +364,10 @@ if (isAppleWinWebKit() || isWinCairo() || isPlayStation() || isFTW()) {
     push @local_options, XcodeCoverageSupportOptions() if $coverageSupport;
     push @local_options, XcodeStaticAnalyzerOption() if $shouldRunStaticAnalyzer;
     push @local_options, "WK_LTO_MODE=$ltoMode" if ($ltoMode ne "default");
+    if (isAppleMacWebKit()) {
+        my $version = osXVersion();
+        push @local_options, "MACOSX_DEPLOYMENT_TARGET=$version->{major}.$version->{minor}";
+    }
```
What about people who use make instead of build-webkit?

Can't we do this in the xcconfig files instead?

emw-apple replied:
Not easily, because (afaik) there isn't a good way to get the `x.y` version number. There's $(MAC_OS_X_VERSION_MINOR), but it's integer formatted (i.e. `MAC_OS_X_VERSION_MINOR = 130300`), and I don't think xcconfig has a way to break that down into a decimal-separated version number.

emw-apple committed:
[Xcode] Set MACOSX_DEPLOYMENT_TARGET on open source builds to match the system version
https://bugs.webkit.org/show_bug.cgi?id=252783

Reviewed by Alexey Proskuryakov.

While internal builds of WebKit do not support back deployment
(targeting an older OS than what an SDK was built for), it's possible to
make open source builds in this way. Currently, we assume that the SDK's
default deployment target is fine, and make no effort to ensure that
what we build is actually runnable on the system.

This feels like unnecessary friction for open source contributors,
especially since the Xcode available in an older OS's App Store contains
SDKs for newer macOS versions. Rather than building artifacts that won't
run on the host machine, we should gracefully fall back.

Achieve this in build-webkit by passing MACOSX_DEPLOYMENT_TARGET, based
on the system's version. The deployment target can be overridden (or
unset) by passing MACOSX_DEPLOYMENT_TARGET on the command line.

See also: https://commits.webkit.org/245182@main, which stopped setting
the *_DEPLOYMENT_TARGET settings automatically in order to support
building for minor-release deployment targets.

* Tools/Scripts/build-webkit:

Canonical link: https://commits.webkit.org/260719@main
